src/hotspot/share/opto/bytecodeInfo.cpp

```diff
@@ -595,7 +604,10 @@ bool InlineTree::ok_to_inline(ciMethod* callee_method, JVMState* jvms, ciCallPro
       set_msg("inline (hot)");
     }
     print_inlining(callee_method, caller_bci, caller_method, true /* success */);
-    build_inline_tree_for_callee(callee_method, jvms, caller_bci);
+    InlineTree* callee_tree = build_inline_tree_for_callee(callee_method, jvms, caller_bci);
+    if (should_delay || AlwaysIncrementalInline) {
+      callee_tree->set_late_inline();
```

Comment:
It took me a while to figure out why this is needed: for replay.  It bothers me a little that AlwaysIncrementalInline is check here and again in the caller.  If the replay file sets should_delay to false, then we shouldn't let AlwaysIncrementalInline to force it to true, right?  So I'm wondering if it would be better to pre-set should_delay to true in the caller if AlwaysIncrementalInline is true.

Reply:
I've added a comment to make it more clear. So, this code is only to record the late inlining decision to later dump it to the replay file. I think initializing `should_delay = AlwaysIncrementalInline` is a good idea. `should_delay` can only become true but not false anymore during normal compilation. 

But I think we need to leave `|| AlwaysIncrementalInline` in here https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/doCall.cpp#L192 in case someone wants to replay compile with that flag even though the replay file recorded a different late inlining decision.

I also fixed the test if run with `-XX:+AlwaysIncrementalInline`.

Reply:
Shouldn't the recorded inlining decision always override flags like -XX:+AlwaysIncrementalInline?
This brings up the question of how to handle flags.  If we stored them in the replay file, then the replay compile could compare those to the current flags and if they don't match:
1) give a warning and continue, correct replay not guaranteed
2) give an error and refuse to continue
3) override current flags with saved flags (this could be implemented by having the "ci" layer cache flags settings for each compile)

Reply:
As we have discussed offline, it's best not to treat `AlwaysIncrementalInline` specially given that we are already enforcing the general inlining decisions based on the replay data. I will therefore remove ` || AlwaysIncrementalInline` from L192.

About the flag handling in general, as you have suggested offline, I also think it's a good option to put the used flags in the replay file in the future and enforce them. If the user wants to run a different set of flags for any reasons then the replay file can be adapted accordingly.